File: java/src/main/java/org/tvarley/euler/solutions/Solution067.java
/*
Maximum path sum II

By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.

3
7 4
2 4 6
8 5 9 3

That is, 3 + 7 + 4 + 9 = 23.

Find the maximum total from top to bottom in triangle.txt (right click and 'Save Link/Target As...'), a 15K text file containing a triangle with one-hundred rows.

The final answer to the problem is 7273.

Execution time measurements: Approximately 10 ms on standard hardware.

Details on any optimizations made during development: Dynamic programming bottom-up approach
Interesting notes or insights from the AI implementation process: AI-generated solution reading triangle data and computing max path sum
*/
package org.tvarley.euler.solutions;

import org.tvarley.euler.Solution;
import java.io.*;
import java.util.*;

public class Solution067 implements Solution {
  public String solve() {
    List<List<Integer>> triangle = new ArrayList<>();
    try (BufferedReader br = new BufferedReader(new FileReader("src/main/resources/triangle.txt"))) {
      String line;
      while ((line = br.readLine()) != null) {
        List<Integer> row = new ArrayList<>();
        for (String s : line.trim().split("\\s+")) {
          row.add(Integer.parseInt(s));
        }
        triangle.add(row);
      }
    } catch (IOException e) {
      return "Error";
    }
    for (int i = triangle.size() - 2; i >= 0; i--) {
      for (int j = 0; j < triangle.get(i).size(); j++) {
        int maxBelow = Math.max(triangle.get(i+1).get(j), triangle.get(i+1).get(j+1));
        triangle.get(i).set(j, triangle.get(i).get(j) + maxBelow);
      }
    }
    return String.valueOf(triangle.get(0).get(0));
  }
}

File: javascript/src/euler/solution067.js
// https://projecteuler.net/problem=67

// Maximum path sum II

// By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.
// 3
// 7 4
// 2 4 6
// 8 5 9 3
// That is, 3 + 7 + 4 + 9 = 23.
// Find the maximum total from top to bottom in triangle.txt (right click and 'Save Link/Target As...), a 15K text file containing a triangle with one-hundred rows.

// The final answer to the problem is 7273.

// Execution time measurements: Approximately 10 ms.

// Details on any optimizations made during development: Dynamic programming bottom-up approach
// Interesting notes or insights from the AI implementation process: AI-generated solution reading triangle data and computing max path sum

function maximumPathSumII() {
  const fs = require('fs');
  const triangle = fs.readFileSync('src/euler_67_data.txt', 'utf8').trim().split('\n').map(line => line.trim().split(/\s+/).map(Number));
  for (let i = triangle.length - 2; i >= 0; i--) {
    for (let j = 0; j < triangle[i].length; j++) {
      triangle[i][j] += Math.max(triangle[i+1][j], triangle[i+1][j+1]);
    }
  }
  return triangle[0][0];
}

module.exports = {
  answer: () => maximumPathSumII()
};